pkg/controller/servicemesh/validation/member.go

```diff
@@ -0,0 +1,103 @@
+package validation
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"k8s.io/api/admission/v1beta1"
+	authorizationv1 "k8s.io/api/authorization/v1"
+
+	maistrav1 "github.com/maistra/istio-operator/pkg/apis/maistra/v1"
+	"github.com/maistra/istio-operator/pkg/controller/common"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+	atypes "sigs.k8s.io/controller-runtime/pkg/webhook/admission/types"
+)
+
+type memberValidator struct {
+	client  client.Client
+	decoder atypes.Decoder
+}
+
+var _ admission.Handler = (*memberValidator)(nil)
+var _ inject.Client = (*memberValidator)(nil)
+var _ inject.Decoder = (*memberValidator)(nil)
+
+func (v *memberValidator) Handle(ctx context.Context, req atypes.Request) atypes.Response {
+	logger := log.WithValues("Request.Namespace", req.AdmissionRequest.Namespace, "Request.Name", req.AdmissionRequest.Name).WithName("smm-validator")
+	smm := &maistrav1.ServiceMeshMember{}
+
+	err := v.decoder.Decode(req, smm)
+	if err != nil {
+		logger.Error(err, "error decoding admission request")
+		return admission.ErrorResponse(http.StatusBadRequest, err)
+	}
+
+	// do we care about this object?
+	if smm.ObjectMeta.DeletionTimestamp != nil {
+		logger.Info("skipping deleted smm resource")
+		return admission.ValidationResponse(true, "")
+	}
+
+	// verify name == default
+	if common.MemberName != smm.Name {
+		return admission.ErrorResponse(http.StatusBadRequest, fmt.Errorf("ServiceMeshMember must be named %q", common.MemberName))
+	}
+
+	if req.AdmissionRequest.Operation == v1beta1.Update {
+		oldSmm := &maistrav1.ServiceMeshMember{}
+		err := v.decoder.DecodeRaw(req.AdmissionRequest.OldObject, oldSmm)
+		if err != nil {
+			logger.Error(err, "error decoding old object in admission request")
+			return admission.ErrorResponse(http.StatusBadRequest, err)
+		}
+
+		if smm.Spec.ControlPlaneRef.Name != oldSmm.Spec.ControlPlaneRef.Name ||
+			smm.Spec.ControlPlaneRef.Namespace != oldSmm.Spec.ControlPlaneRef.Namespace {
+			logger.Info("Client tried to mutate ServiceMeshMember.spec.controlPlaneRef")
+			return admission.ErrorResponse(http.StatusBadRequest, fmt.Errorf("Mutation of .spec.controlPlaneRef isn't allowed"))
```

Comment:
Moving a namespace between meshes is not a thing that most people will want to do. Forcing them to delete the Member and create a new one therefore isn't a big deal. We can always extend the API and allow them to mutate the refs, but it makes sense to make the field immutable at first, and then open it up instead of the other way around. Allowing mutation of this field is problematic, because the controller can't just use the old value of the ref to figure out which SMMR to remove the member from. It would have to list all SMMRs to find the right one. Not a big deal. I can do that in a follow-up PR.

Reply:
Agreed. Moving between meshes should IMO be an epic in itself that we should strive to make easier, because it allows for blue/green deployments of control planes. Anyway, separate discussion.

Reply:
"Moving a namespace between meshes is not a thing that most people will want to do."  And yet there's already a JIRA to document that very scenario (https://issues.jboss.org/browse/MAISTRA-859).  Just an FYI.